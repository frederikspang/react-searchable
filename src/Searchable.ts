import {ChangeEvent, Component, ReactNode} from 'react';

type HandleChange = (event: ChangeEvent<HTMLInputElement>) => void;

interface IProps<T> {
  initialValue?: string;
  items: T[];
  predicate(item: T, value: string): boolean;
  children(params: {
    items: T[];
    handleChange: HandleChange;
    value: string;
  }): ReactNode;
}

interface IState<T> {
  items: T[];
  value: string;
}

export default class Searchable<T> extends Component<IProps<T>, IState<T>> {
  public state: IState<T> = {
    items: this.props.items,
    value: this.props.initialValue || '',
  };

  public handleChange: HandleChange = ({target: {value}}) =>
    this.setState({value})

  public componentDidUpdate(prevProps: IProps<T>, prevState: IState<T>): void {
    const {value} = this.state;

    if (value !== prevState.value) {
      const {items, predicate} = this.props;

      this.setState({items: items.filter((item) => predicate(item, value))});
    }
  }

  public render(): ReactNode {
    const {items, value} = this.state;

    return this.props.children({items, value, handleChange: this.handleChange});
  }
}
