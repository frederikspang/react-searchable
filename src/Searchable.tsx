import debounce from 'lodash.debounce';
import {ChangeEvent, Component, ReactNode} from 'react';

type HandleChange = (event: ChangeEvent<HTMLInputElement>) => void;

export interface IProps<T> {
  debounceDuration: number;
  initialValue: string;
  items: T[];
  predicate(item: T, value: string): boolean;
  children(params: {
    items: T[];
    handleChange: HandleChange;
    value: string;
  }): ReactNode;
}

interface IState<T> {
  items: T[];
  value: string;
}

export default class Searchable<T> extends Component<IProps<T>, IState<T>> {
  public static defaultProps = {
    debounceDuration: 100,
    initialValue: '',
  };

  constructor(props: IProps<T>) {
    super(props);

    const { initialValue: value, predicate, items, debounceDuration } = props;

    this.state = {
      items: value !== '' ? this.filter(items, value) : items,
      value,
    };

    this.filterAndSetState = debounce(this.filterAndSetState, debounceDuration);
  }

  public filterAndSetState(items: T[], value: string): void {
    this.setState({
      items: value !== '' ? this.filter(items, value) : items,
    });
  }

  public filter = (items: T[], value: string): T[] => items.filter((item) => this.props.predicate(item, value));

  public handleChange: HandleChange = ({target: {value}}) =>
    this.setState({value})

  public componentDidUpdate(prevProps: IProps<T>, prevState: IState<T>): void {
    const {value} = this.state;

    if (value !== prevState.value) {
      const {items} = this.props;

      this.filterAndSetState(items, value);
    }
  }

  public render(): ReactNode {
    const {items, value} = this.state;
    const {children} = this.props;

    return children({items, value, handleChange: this.handleChange});
  }
}
