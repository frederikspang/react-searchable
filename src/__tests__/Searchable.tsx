import {shallow, ShallowWrapper} from 'enzyme';
import * as React from 'react';
import Searchable, {IProps} from '../Searchable';

jest.mock('lodash.debounce', (): any => jest.fn((fn: () => any) => fn));

interface IUser {
  name: string;
}

interface ISetup<T> {
  wrapper: ShallowWrapper;
  props: IProps<T>;
}

function setup(overrides?): ISetup<IUser> {
  const props: IProps<IUser> = {
    items: [
      {
        name: 'Stewart Harrison',
      },
      {
        name: 'Jake Bullock',
      },
    ],
    predicate: (item: IUser, value: string): boolean =>
      item.name.includes(value),
    ...overrides,
  };

  const wrapper: ShallowWrapper = shallow(
    <Searchable {...props}>{() => <></>}</Searchable>,
  );

  return {
    props,
    wrapper,
  };
}

describe('<Searchable />', () => {
  it('renders without crashing', () => {
    const {wrapper} = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('filters initial state based on props', () => {
    const {wrapper} = setup({initialValue: 'Jake'});

    expect(wrapper.state('items')).toEqual([{name: 'Jake Bullock'}]);
  });

  it('returns empty array if no matches', () => {
    const {wrapper} = setup();

    wrapper.setState({value: 'corncob'});

    expect(wrapper.state('items')).toEqual([]);
  });

  it('filters items based on predicate', () => {
    const {wrapper} = setup();

    wrapper.setState({value: 'Jake'});

    expect(wrapper.state('items')).toEqual([{name: 'Jake Bullock'}]);
  });

  describe('handleChange', () => {
    it('sets state', () => {
      const {wrapper} = setup();
      const instance = wrapper.instance() as Searchable<IUser>;
      const event = { target: { value: 'Jake' } } as React.ChangeEvent<HTMLInputElement>;

      instance.handleChange(event);

      expect(wrapper.state('value')).toEqual('Jake');
    });
  });
});
