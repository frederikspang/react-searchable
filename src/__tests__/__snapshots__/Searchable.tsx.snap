// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Searchable /> renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Searchable
    debounceDuration={100}
    initialValue=""
    items={
      Array [
        Object {
          "name": "Stewart Harrison",
        },
        Object {
          "name": "Jake Bullock",
        },
      ]
    }
    predicate={[Function]}
  >
    [Function]
  </Searchable>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
